### PR TITLE
Add FAISS DocumentStore integration

### LINKED ISSUE #717: [Integration] Request to support FAISSDocumentStore in 2.x
**Is your feature request related to a problem? Please describe.**
We are having problem migrating current version to 2.0 due to no FAISSSupport
**Describe the solution you'd like**
Due to is simplicity , performance and developer experience , we and many other RAG solution provider are now using FAISS . So it would be nice to integrate it.

**Describe alternatives you've considered**
QDrant and many others are not approved by some organization yet because they are very new and not backed by big vendors.

### PR DESCRIPTION
**Pull request by GunaPalanivel**
## Description

- Fixes #717

Adds a new **FAISSDocumentStore** integration for Haystack 2.x.

### Core Implementation

* Introduced `FAISSDocumentStore` with full Haystack 2.x protocol support

### Features

* In-memory and persistent storage (FAISS index + JSON metadata)
* Configurable FAISS index types
* Vector similarity search
* Comprehensive metadata filtering (comparison + logical operators)
* Full document CRUD support

### Persistence

* FAISS index for vector storage
* In-memory metadata dictionary
* JSON serialization for metadata persistence

### Error Handling

* Proper validation with `FilterError` for invalid filter operations

## Testing

* 55 tests passing, 2 skipped
* Uses Haystack’s standard DocumentStore test mixins
* Covers:

  * Filtering logic
  * Persistence
  * Metadata operations
  * Edge cases

Run:

```bash
hatch run test:all
# 55 passed, 2 skipped
```

## Checklist

* [x] Follows Haystack 2.x conventions
* [x] Comprehensive test coverage (standard mixins)
* [x] Formatted and linted (ruff)
* [x] Persistence verified
* [x] Metadata filtering validated
* [x] Documentation updated (docstrings + README)

### CODE REVIEW

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+        sql_url: str = "sqlite:///:memory:",  # Kept for compatibility but unused
+        index_path: str | None = None,
+        index_string: str = "Flat",
+        embedding_dim: int = 768,
+        faiss_index_factory_str: str | None = None,  # Legacy parameter support
+        similarity: str = "dot_product",  # Legacy parameter
+        isolation_level: str | None = None,  # Legacy parameter
+        duplicate_documents: str = "overwrite",  # Legacy parameter
+        return_embedding: bool = True,
+        progress_bar: bool = True,  # Legacy parameter
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py]: Seems that only those attributes are actually used, and even serialised:

```python
        index_path: str | None = None,
        index_string: str = "Flat",
        embedding_dim: int = 768,
```

Can you remove all the others? Or what's the idea behind them?

Comment (GunaPalanivel) [reply integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py]: Removed unused legacy parameters (a7d6570)

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
+import json
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py:5]: Add the LICENSE here

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+            )  # Simplified score handling
+
+            results.append(result_doc)
+
+            if len(results) >= top_k:
+                break
+
+        return results
+
+    def _get_result_to_documents(self, result) -> list[Document]:
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py]: This is currently not being used anywhere, we can remove it

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+        docs_to_delete = self.filter_documents(filters)
+        ids = [doc.id for doc in docs_to_delete]
+        self.delete_documents(ids)
+        return len(ids)
+
+    def count_documents_by_filter(self, filters: dict[str, Any]) -> int:
+        return len(self.filter_documents(filters))
+
+    def update_by_filter(self, filters: dict[str, Any], meta: dict[str, Any]) -> int:
+        docs_to_update = self.filter_documents(filters)
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py:399]: Missing the docstring here, please make sure to mention that the updates are in-memory only.

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+                        type_name = "keyword"
+                    elif type_name == "int":
+                        type_name = "long"
+                    elif type_name == "bool":
+                        type_name = "boolean"
+                    fields_idx[key] = {"type": type_name}
+        return fields_idx
+
+    def get_metadata_field_min_max(self, field_name: str) -> dict[str, Any]:
+        values = []
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py:433]: We need to add a docstring here.

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+            if val is not None:
+                values.append(val)
+
+        if not values:
+            return {"min": None, "max": None}
+
+        return {"min": min(values), "max": max(values)}
+
+    def get_metadata_field_unique_values(self, field_name: str) -> list[Any]:
+        values = set()
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py:455]: We need to add a docstring here.

File: integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+                self._get_doc_value(doc, field_name)
+                if not field_name.startswith("meta.")
+                else doc.meta.get(field_name[5:])
+            )
+            if val is not None:
+                values.add(val)
+        return list(values)
+
+    def count_unique_metadata_by_filter(self, filters: dict[str, Any], fields: list[str]) -> dict[str, int]:
+        filtered_docs = self.filter_documents(filters)
```
Comment (davidsbatista) [inline integrations/faiss/src/haystack_integrations/document_stores/faiss/document_store.py:474]: We need to add a docstring here.

File: integrations/faiss/tests/test_document_store.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import pytest
```
Comment (davidsbatista) [inline integrations/faiss/tests/test_document_store.py:5]: There are few tests which are still not part of our Mixin, so we need to add them for each doc store:

- count_documents_by_filter
- get_metadata_fields_info
- count_unique_metadata_by_filter
- search(query_embedding, top_k)
- search with filters


Also, we need to add tests for:
- to_dict / from_dict with round-trip
- Tests for load when .faiss or .json is missing, and for save/load when some documents have no embeddings

Comment (davidsbatista) [inline integrations/faiss/tests/test_document_store.py:5]: Also add the LICENSE here

**Comment by GunaPalanivel** [comment]
> Thanks again for this @GunaPalanivel, and sorry for the delay. I did another review round, a few more things to fix/improve.

Thanks for the thorough review @davidsbatista

Here’s what’s been addressed in this round:

* Added SPDX license headers to both `document_store.py` and `test_document_store.py`.
* Removed the unused `_get_result_to_documents` method.
* Added docstrings to:

  * `update_by_filter` (including note on in-memory behavior),
  * `delete_by_filter`,
  * `count_documents_by_filter`,
  * `get_metadata_fields_info`.
* Added the requested test coverage for:

  * `count_documents_by_filter`
  * `get_metadata_fields_info`
  * `count_unique_metadata_by_filter`
  * search (with and without filters)
  * `to_dict` / `from_dict` roundtrip
  * persistence (with and without embeddings)
  * loading with missing files

All tests are passing locally. Please let me know if anything else needs tightening.

**Comment by GunaPalanivel** [comment]
> > Thanks @GunaPalanivel for the changes. In order for this Documentstore to be used with Haystack there's one essential component missing, an EmbeddingRetriever.
> > We need to add a `FAISSEmbeddingRetriever` that accepts a `query_embedding` (and optional filters, top_k) and calls document_store.search(query_embedding, top_k, filters=filters).
> 
> Thanks for the clear feedback, @davidsbatista! Understood — I'll add a `FAISSEmbeddingRetriever` as a proper `@component`-decorated class with:
> 
> * `run(query_embedding, filters, top_k)` calling `document_store.search(...)`
> * `FilterPolicy` support with `apply_filter_policy`
> * `to_dict` / `from_dict` for pipeline serialization
> * Tests and updated `__init__.py` exports
> 
> Will push the changes shortly!

Added `FAISSEmbeddingRetriever` in `components/retrievers/faiss/embedding_retriever.py`:

- `@component`-decorated with `run()`, `run_async()`, `to_dict()`, `from_dict()`
- `FilterPolicy` support via `apply_filter_policy`
- Full pipeline serialization with backward-compat `filter_policy` guard
- 8 new tests (72 total, 0 regressions)
- `pyproject.toml` mypy now covers both sub-packages

Ready for re-review @davidsbatista
